include/gul14/ThreadPool.h

```diff
@@ -0,0 +1,297 @@
+ * stored in a queue and executed in the order they were added. This makes a ThreadPool
+ * with a single thread effectively a serial task queue.
+ *
+ * Each task is associated with a unique ID. This ID is returned by add_task() and can
+ * be used to query the status of the task or to remove it from the queue via
+ * remove_pending_task(). Tasks that are already running cannot be canceled.
+ *
+ * Tasks can be scheduled to start at a specific time point or after a certain delay. Each
+ * task can also be given a name, which is mainly useful for debugging.
+ *
+ * \code{.cpp}
+ * ThreadPool pool(2); // Create a pool with 2 threads
+ *
+ * pool.add_task([]() { std::cout << "Task 1\n"; });
+ * pool.add_task([]() { sleep(1); std::cout << "Task 2\n"; });
+ *
+ * // Start 2 seconds after enqueueing (if a thread is available)
+ * pool.add_task([]() { std::cout << "Task 3\n"; }, 2s);
+ *
+ * // Probable output:
+ * // Task 1
+ * // Task 2
+ * // Task 3
+ *
+ * sleep(3); // Wait for the tasks above to finish
+ *
+ * // Tasks can also interact with the pool themselves, e.g. to schedule a continuation:
+ * pool.add_task(
+ *     [](ThreadPool& pool) {
+ *         std::cout << "Task 4\n";
+ *         pool.add_task([]() { std::cout << "Task 5, a second later\n"; }, 1s);
+ *     });
+ * \endcode
+ *
+ * All public member functions are thread-safe.
+ *
+ * On Linux, threads in the pool explicitly block the signals SIGALRM, SIGINT, SIGPIPE,
+ * SIGTERM, SIGURG, SIGUSR1, and SIGUSR2. This is done to prevent the threads from
+ * terminating the whole process if one of these signals is received.
+ *
+ * \since GUL version 2.11
+ */
+class ThreadPool
+{
+public:
+    using TaskId = std::uint64_t;
+    using TimePoint = std::chrono::time_point<std::chrono::system_clock>;
+    using Duration = TimePoint::duration;
+
+    /// Maximum possible capacity for the task queue.
+    constexpr static std::size_t max_capacity{ 10'000'000 };
+
+    /// Maximum possible number of threads
+    constexpr static std::size_t max_threads{ 10'000 };
+
+    /**
+     * Create a thread pool with the desired number of threads and the specified capacity
+     * for queuing tasks.
+     *
+     * This constructor launches the desired number of threads. The threads are joined
+     * when the ThreadPool object gets destroyed.
+     *
+     * \param num_threads  Desired number of threads
+     * \param capacity     Maximum number of pending tasks that can be queued
+     *
+     * \exception std::invalid_argument is thrown if the desired number of threads is
+     *            zero or greater than max_threads, or if the requested capacity is zero
+     *            or exceeds max_capacity.
+     */
+    ThreadPool(std::size_t num_threads, std::size_t capacity = 1000);
+
+    /**
+     * Destruct the ThreadPool and join all threads.
+     *
+     * This destructor blocks until all threads have terminated. Work that has not yet
```

Comment:
>     * Some kind of signal to the running worker would be nice "finish your work now!"

Hm... we can already pass a `ThreadPool&` to the task function. We could provide a public function `bool is_cancel_requested()` or similar. The task could check that and terminate itself. Does that make sense?

>     * Would also be nice if we have a function that force-stops the threads.

I wouldn't know how, at least not without going down to platform-specific APIs (pthread cancel or such) and potentially leaking resources. If the thread is somehow force-stopped, C++ destructors are not run and we are in a special kind of hell. Hm... one thing we *could*  do with standard threads is to `detach()` them. This way they continue to run (or maybe end) and we do not have to care about them anymore. But it has a smell...

Reply:
> Hm... we can already pass a `ThreadPool&` to the task function. We could provide a public function `bool is_cancel_requested()` or similar. The task could check that and terminate itself. Does that make sense?

No ;-) The ref is either valid (when we can continue) or dangling (then we should never ever access it) when we should stop :-}

We would need a `weak_ptr` to a `shared_ptr` to the `ThreadPool`, and the `shared_ptr` gets freed on `~ThreadPool`, so that we can witness that in the `weak_ptr`. Welcome `safe_self` ;-)

Reply:
> platform-specific APIs (pthread cancel or such)

Don't we do that already anyhow with the signal blocking?

Reply:
Ok, destructors.

But at least we need some handle-ish thing were the task can detect it shall stop. We do not have that. Some kind of condition variable would be even better.

Reply:
```c++
class ThreadPool {
    std::shared_ptr<ThreadPool> safe_self_;
    ...
    // call the working task with
    std::weak_ptr<ThreadPool> x = safe_self_;
    worker(x);

    ThreadPool() {
        safe_self_ = this;
    }

    ~ThreadPool {
        safe_self_.reset(); // do this first
        // further destructive work follows
    }
}

// In the worker loop
worker(std::weak_ptr<ThreadPool> tp) {
    while (not tp.expired()) {
        // do work
    }
}
```

_Edit: Forgot **not**_
_Edit: Add ctor and dtor example snippets_

Reply:
Hmmm... I am trying to use this "safe-self" idiom for the new `TaskHandle`, but I have trouble understanding it. If the user allocates the `ThreadPool` with `make_shared()`, everything is fine, of course, because it can be kept alive as long as it is needed. But we allow simple construction on the stack:
```cpp
TaskHandle task; // a bit contrived and not directly possible, but still...
{
    ThreadPool pool;
    task = pool.add_task([](){ ... });
} // (1)
```
Now, nothing is stopping `~ThreadPool()` from running at (1). Sure, the worker can detect that the pool has `expired()` by querying the weak pointer at any point...
```cpp
if (not weak_pointer.expired()) // (2)
    do_something(); // (3) 
```
But unfortunately that does not save us because nothing guarantees that the thread pool has not been destructed between (2) and (3). Right?

Reply:
I guess this could work if the `ThreadPool` was just a façade over an internal `ThreadPoolEngine` that is allocated via `make_shared()`, though. Maybe that is the way to go?

Reply:
The entire `ThreadPool` is now re-engineered so that `add_task()` hands out a `TaskHandle` and that maintains a safe reference to its parent pool (via weak_ptr). If tasks want to, they can accept a `ThreadPoolEngine&` reference and ask it if `is_shutdown_requested()`.

Reply:
Regarding your comment 2 up:
> _But unfortunately that does not save us because nothing guarantees that the thread pool has not been destructed between (2) and (3). Right?_

That depends on the goal. I must admit I have lost track.

What it **does** solve is a communication channel from the ThreadPool to the Task 'hey task, if it is convenient for you please stop'. As far as I remember there is no condition variable or anything that communicates the stop request to the Task. The Task can also not ask the ThreadPool because that might have died in between. The step (3) above is not calling the ThreadPool. The whole setup is just to make the Task stop if the ThreadPool wants to die (and it can only die after all Tasks give also up).

I have not read your recent changes (which sounded excellent), but at least that is a way to say: "Please please Task, if you see your weak-ptr gone, I really need to die... let me go!"

Ah, 

> because nothing guarantees that the thread pool has not been destructed between (2) and (3)

Of course there is, as _you_ pointed out. the ThreadPool dtor waits for the Task to quit, so that it can join. It will halt destruction until the Task is gone.

What was missing was just the comm channel to inform the Task that it should maybe quit now and not troddle on with all the looping, because ... well, someone wants all Tasks to stop.

I think this might not be needed
> _façade over an internal ThreadPoolEngine that is allocated via_

Which problem is not solved by the 'safe-self' thingy?  
Of course the communication can also be with several other means, like a special selfmade object, or a condition variable. But then, maybe the safe-self is just as easy to implement and less overhead and/or less resources.

Of course, at least how I wrote it above, you must not call the weak-ptr. Well for that you would have `weak_pointer.lock()`, which should also all be good, because the ThreadPool dtor is waiting for us anyhow.

Reply:
And after reading the docs on `shared_ptr::reset()`
> If *this already owns an object and it is the last shared_ptr owning it, the object is destroyed through the owned deleter.

The code details above is too simple. I guess you need a data member thats a regular pointer to self and a shared_ptr that holds a reference to THAT member data thing instead to the actual object. Sounds complicated, but is not ;) Lets have a look at my actual code...

```c++
class D_ZMQ  {
  private:
       D_ZMQ*                   self_{this};
  protected:
       std::shared_ptr<D_ZMQ*>  safe_self_;
...
  public:
       D_ZMQ::D_ZMQ (const std::string &pn, EqFct *ef, bool sender)
       {
           safe_self_ = std::make_shared<D_ZMQ*>(self_); // note 1
           ...
       }

       D_ZMQ::~D_ZMQ (void)
       {
           safe_self_ = std::make_shared<D_ZMQ*>(nullptr); // note 2
       }
};
```
**Note 1:** I think it was impossible for some reason to do it in the member initializer
**Note 2:** This is the same as `safe_self_.reset()` _I guess_ :grimacing: 

Handing out here is a bit complicated, but its just passed to a function that takes a weak-ptr which transmogrifies it
```c++
void D_ZMQ::register_MPchecker()
{
    ... 
    MPchecker_.push_back(MPchecker_idx_, prop, safe_self_); // safe_self_ taken as weak_ptr
}
```

Here is the other object that receives the safe-self:
```c++
void MacroPulseChecker::push_back(int& index, std::string prop_name, std::weak_ptr<doocs::D_ZMQ*> prop_ptr)
{
    ...
    entry.prop_ptr = std::move(prop_ptr);
}
```

And here it is just used to remove registered properties from the MPchecker if the property got destructed in the meantime and did not / could not inform the MPchecker to remove it

```c++
std::string MacroPulseChecker::get_fill_status(int bufno)
{
    ...
    std::string ret =
        std::accumulate(elem.begin(), elem.end(), std::string{},
            [&at_least_one_filled, this] (auto const& a, auto const& e) {
            ...
            if (e.prop_ptr.expired())
                this->remove(idx); // remove removed properties
            return a + std::to_string(idx) + ", "s;
            });
    ...
}
```

`e.prop_ptr` is not called, it is just used as signal the prop has destructed meanwhiles.

Reply:
After having skimmed through the current diff and with all knowledge gained ;-)

I guess this is not needed, as it is the same as, in the Task

```diff
-if (not weak_pointer.expired()) // (2)
-    do_something(); // (3) 
+if (not treadpool_reference.is_shutdown_requested()) // (2)
+    do_something(); // (3) 
```

And again I need some thought if deadlocks are possible, because you use this one-mutex-for-multiple-variables instead of just making them atomic? Not sure if you need the longer lived lock somewhere, but for the above signalling flag thingy it is certainly more simple to just use an atomic?

So I close this here :tada: